### ISSUE TITLE
Add label argument to `@defer` to support Relay

### ISSUE THREAD

**Issue by levibuzolic** [rmosolgo/graphql-ruby#3454]
Relay 11 has added initial support for `@defer` however it requires support for a `label` argument which is missing in the current version of GraphQL Pro. The defer RFC spec outlines the requirements here: https://github.com/graphql/graphql-spec/blob/main/rfcs/DeferStream.md

Here's a patch file that I beleive adds the neccesary requirements, ~however (potentially unrelated) I haven't been able to get this working locally yet as the `self.resolve` doesn't seem to be getting called on the defer directive.~ I'll keep investigating this on my own, but in the meantime here's the patch for `defer.rb`:

```diff
--- defer.rb	2021-04-28 14:28:27.000000000 +1000
+++ defer.rb	2021-04-28 14:46:05.000000000 +1000
@@ -16,11 +16,14 @@
         default_value: true,
         description: "If false, this field _won't_ be deferred."
 
+      argument :label, String, required: false,
+        description: "A unique label to identify the deferred payload."
+
       def self.resolve(object, arguments, context, &block)
         if arguments[:if]
           defer = context[:defer] ||= Deferred.new(context)
           path = context.namespace(:interpreter).fetch(:current_path)
-          defer.deferrals << Deferral.new(block: block, path: path, context: context)
+          defer.deferrals << Deferral.new(block: block, path: path, context: context, label: arguments[:label])
         else
           yield
         end
@@ -69,12 +72,16 @@
         # @return [Array<String, Integer>] The path to this deferred part of the query
         attr_reader :path
 
+        # @return [String, nil] A unique label to identify the deferred payload
+        attr_reader :label
+
-        def initialize(path:, block:, context:)
+        def initialize(path:, block:, context:, label:)
           # This path was a new object created by the runtime,
           # so it doesn't need duping.
           @path = path
           @block = block
           @context = context
+          @label = label
           @data = nil
           @errors = nil
           @block_was_run = false
@@ -98,6 +105,10 @@
             path: path,
           }
 
+          if label
+            h[:label] = label
+          end
+
           if errors
             h[:errors] = errors
           else
```

**Comment by rmosolgo** [comment]
Thanks for the feature request and the proposed patch! 

Are you using `@defer` with Relay? If so, that might be why you aren't seeing `self.resolve` getting called. My understanding is that Relay "compiles away" `@defer` and sends several different queries to fulfill the one that had `@defer` in it. (It's a bit hard to tell though ... the Relay docs link to some internal Facebook website 😖 ) 

I'll take a look at adding this argument and follow up here soon.

**Comment by levibuzolic** [comment]
@rmosolgo yeah the lack of documentation is really frustrating. I'm trying to work through this step by step to see what problems needs to be solved to get this working, but it may still not be quite ready for public consumption. I still haven't cracked how to properly define the JS networking stack for Relay to handle streamed responses, there seems to be some API there but I haven't worked out how it hangs together yet.

I'm reasonably sure the `@defer` directives aren't compiled away anymore as of Realy 11. Here's a sample generated file where you can see Relay is automatically generating a label for the directive on a fragment, along with an unlabled `@defer` directive on a scalar field, both of which are being received successfully on the server:

```js
/**
 * @flow
 */

/* eslint-disable */

'use strict';

/*::
import type { ConcreteRequest } from 'relay-runtime';
import type { Fragment_card$ref } from "./Fragment_card.graphql";
export type ExampleQueryVariables = {||};
export type ExampleQueryResponse = {|
  +decks: $ReadOnlyArray<{|
    +id: string,
    +name: ?string,
    +slots: $ReadOnlyArray<{|
      +id: string,
      +quantity: ?number,
      +card: ?{|
        +$fragmentRefs: Fragment_card$ref
      |},
    |}>,
  |}>
|};
export type ExampleQuery = {|
  variables: ExampleQueryVariables,
  response: ExampleQueryResponse,
|};
*/


/*
query ExampleQuery {
  decks {
    id
    name @defer
    slots {
      id
      quantity
      card {
        ...Fragment_card @defer(label: "ExampleQuery$defer$Fragment_card")
        id
      }
    }
  }
}

fragment Fragment_card on Card {
  id
  name
  type
  price
}
*/

const node/*: ConcreteRequest*/ = (function(){
var v0 = {
  "alias": null,
  "args": null,
  "kind": "ScalarField",
  "name": "id",
  "storageKey": null
},
v1 = {
  "alias": null,
  "args": null,
  "kind": "ScalarField",
  "name": "name",
  "storageKey": null
},
v2 = {
  "alias": null,
  "args": null,
  "kind": "ScalarField",
  "name": "quantity",
  "storageKey": null
};
return {
  "fragment": {
    "argumentDefinitions": [],
    "kind": "Fragment",
    "metadata": null,
    "name": "ExampleQuery",
    "selections": [
      {
        "alias": null,
        "args": null,
        "concreteType": "Deck",
        "kind": "LinkedField",
        "name": "decks",
        "plural": true,
        "selections": [
          (v0/*: any*/),
          (v1/*: any*/),
          {
            "alias": null,
            "args": null,
            "concreteType": "Slot",
            "kind": "LinkedField",
            "name": "slots",
            "plural": true,
            "selections": [
              (v0/*: any*/),
              (v2/*: any*/),
              {
                "alias": null,
                "args": null,
                "concreteType": "Card",
                "kind": "LinkedField",
                "name": "card",
                "plural": false,
                "selections": [
                  {
                    "kind": "Defer",
                    "selections": [
                      {
                        "args": null,
                        "kind": "FragmentSpread",
                        "name": "Fragment_card"
                      }
                    ]
                  }
                ],
                "storageKey": null
              }
            ],
            "storageKey": null
          }
        ],
        "storageKey": null
      }
    ],
    "type": "Query",
    "abstractKey": null
  },
  "kind": "Request",
  "operation": {
    "argumentDefinitions": [],
    "kind": "Operation",
    "name": "ExampleQuery",
    "selections": [
      {
        "alias": null,
        "args": null,
        "concreteType": "Deck",
        "kind": "LinkedField",
        "name": "decks",
        "plural": true,
        "selections": [
          (v0/*: any*/),
          (v1/*: any*/),
          {
            "alias": null,
            "args": null,
            "concreteType": "Slot",
            "kind": "LinkedField",
            "name": "slots",
            "plural": true,
            "selections": [
              (v0/*: any*/),
              (v2/*: any*/),
              {
                "alias": null,
                "args": null,
                "concreteType": "Card",
                "kind": "LinkedField",
                "name": "card",
                "plural": false,
                "selections": [
                  (v0/*: any*/),
                  {
                    "if": null,
                    "kind": "Defer",
                    "label": "ExampleQuery$defer$Fragment_card",
                    "selections": [
                      (v0/*: any*/),
                      (v1/*: any*/),
                      {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "type",
                        "storageKey": null
                      },
                      {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "price",
                        "storageKey": null
                      }
                    ]
                  }
                ],
                "storageKey": null
              }
            ],
            "storageKey": null
          }
        ],
        "storageKey": null
      }
    ]
  },
  "params": {
    "cacheID": "90ccc89a2a6fc0faf895340120ad9145",
    "id": null,
    "metadata": {},
    "name": "ExampleQuery",
    "operationKind": "query",
    "text": "query ExampleQuery {\n  decks {\n    id\n    name @defer\n    slots {\n      id\n      quantity\n      card {\n        ...Fragment_card @defer(label: \"ExampleQuery$defer$Fragment_card\")\n        id\n      }\n    }\n  }\n}\n\nfragment Fragment_card on Card {\n  id\n  name\n  type\n  price\n}\n"
  }
};
})();
// prettier-ignore
(node/*: any*/).hash = '325cf223d810e19faf9e2ed6955f3a27';

module.exports = node;
```

**Comment by rmosolgo** [comment]
I just shipped graphql-pro v1.17.15 with a patch like the one you suggested above. Please let me know if you run into any trouble with `label:`! 

Meanwhile... 

> I'm seeing the `Schema.execute` call block until everything is complete, rather than streaming its output. 

I'm surprised to hear that! Here's a minimal example of how it's _supposed_ to work: 

<details>
<summary>Deferrals run one-at-a-time</summary>

<p>


```ruby 
require 'bundler/inline'

gemfile do
  gem 'graphql'
  gem 'graphql-pro'
end

class Query < GraphQL::Schema::Object
  field :add, Integer, null: false do
    argument :left, Integer, required: true
    argument :right, Integer, required: true
  end

  def add(left:, right:)
    puts "- Execute: #{left} + #{right}"
    (left || 100) + (right || 1000)
  end
end

class Schema < GraphQL::Schema
  query(Query)
  use GraphQL::Pro::Defer
end

query_str = <<-GRAPHQL
{
  one: add(left: 1, right: 0)
  two: add(left: 2, right: 0) @defer
  three: add(left: 3, right: 0) @defer
}
GRAPHQL

puts "Schema.execute:"
result = Schema.execute(query_str)
puts "Call each deferral:"
result.context[:defer].each.each_with_index do |deferral, index|
  puts "- Deferral #{index}: #{deferral.to_h}"
end
```

</p>

</details>

It _should_ print out alternating messages from execution and result handling, showing that control flow is passing between the two:

```
$ ruby defer_test.rb
Schema.execute:
- Execute: 1 + 0
Call each deferral:
- Deferral 0: {:data=>{"one"=>1}}
- Execute: 2 + 0
- Deferral 1: {:path=>["two"], :data=>2}
- Execute: 3 + 0
- Deferral 2: {:path=>["three"], :data=>3}
```

**Comment by fazo96** [comment]
I am also trying to use `@defer` with Relay 11 with poor results unfortunately.

I am on graphql-pro 1.17.15 and the `@defer` keyword only works on scalar fields, and won't work on arrays or fragment expansions.

According to people on the `#relay` channel on the GraphQL slack, Relay only supports `@defer` on fragment expansions, for example: `...FreelancerCardSkillsInfo_Freelancer @defer(label: "Skills")`

However, that won't work on graphql-ruby because this `@defer` keywork is ignored by the server, which just returns the data normally, then Relay ignores it unfortunately.

@rmosolgo it would be a great help if we could get another patch to make `@defer` work on fragment expansions like in the example above :)

**Comment by rmosolgo** [comment]
Hey @fazo96 , I'd be happy to look into it. Would you mind opening a new issue for it? If you could share a few example queries, along with what you _expected_ to happen and what _actually_ happened, that'd help, too.

**Comment by fazo96** [comment]
Hi @rmosolgo , thank you for the quick response! I have opened a separate issue https://github.com/rmosolgo/graphql-ruby/issues/3495

**Comment by rmosolgo** [comment]
👋  I just released graphql v1.12.12 which improves how runtime directives are applied to inline fragments and fragment spreads (er, it applies them _at all_). Please give it a shot on that new version and let me know if you still run into trouble!
